### ISSUE TITLE
pure attention mode

### ISSUE THREAD

**Issue by sw005320** [espnet/espnet#1318]
I think pure attention/ctc mode is not well documented. We just have https://github.com/espnet/espnet#ctc-attention-and-hybrid-ctcattention but this could not be working because of the lack of appropriate setting of several hyperparameters including 
```
maxlenratio: 0.0
minlenratio: 0.0
```
should be
```
maxlenratio: 0.8
minlenratio: 0.3
```
or whatever.

Probably we should put some warning or chaning the default values of `maxlenratio` and `minlenratio` when `ctc-weight: 0.0` (?)
Any thoughts?

**Comment by Ajyy** [comment]
Hi,

We have already follow your suggestion to set the parameters

`maxlenratio: 0.8
minlenratio: 0.3`

in both training and testing of a pure attention system.
The result is still extremely bad.

![543ec11501a8e7a784fcdc484928d11](https://user-images.githubusercontent.com/38291486/67827683-a7cdb880-fb0b-11e9-9f25-65f00f9f5d3e.png)

And by checking the decoding log, we find that the sentence length predicted may be much larger than the length of groundtruth like the above screenshot shown.

We are working on the librispeech recipe, we are using a smaller model and reduce the training data to 100 hours.
We are using 5000 modeling unit generated by BPE.

Could you tell us if there are any other setting we should be aware of ?

On the other hand, we are also getting very bad result on a pure CTC system.
We have already set the ctc-weight to 1.0 in both training and testing.

In fact, only when we are using a mix system( mtlalpha=0.3), the result is normal.


Thanks a lot!

**Comment by sw005320** [comment]
This is the difficult part of the attention handling.
You can reduce the maxlenratio and minlenratio depending on the language or BPE unit.
You can also control the insertion penalty.
These control the length of the hypotheses, and probably you'll get some reasonable numbers when you reduce both ratios or increase the penalty.
Another possibility is to implement the coverage term during the beam search (instead we adopt to use CTC).

> On the other hand, we are also getting very bad result on a pure CTC system. We have already set the ctc-weight to 1.0 in both training and testing.

Could you also show us the result?
My guess is `ctc_beam`, but not sure.

@ShigekiKarita, could you check the pure CTC decoding mode in the transformer setup? 
You seem to correctly disable `ctc_beam` but I'm not completely sure.
https://github.com/espnet/espnet/search?q=ctc_beam&unscoped_q=ctc_beam

**Comment by Ajyy** [comment]
The CTC result is:

![2019-10-31 10-05-59 的屏幕截图](https://user-images.githubusercontent.com/38291486/67913713-eb3a2c80-fbc8-11e9-8a7c-8f2dd56e809f.png)

Back to the pure attention mode, do I need to set `minlenratio, maxlenration, ctc_weight and penalty` for the training part? Because I find these four parameters also appear in the asr_train.py.
(Although the result is also bad when setting them at both training and decoding part)

> You can reduce the maxlenratio and minlenratio depending on the language or BPE unit.
You can also control the insertion penalty.

With your suggestion, I reduce the value of these two parameters as 0.1 and 0.05 and set the penalty to 10. The length of the sentence predicted is much shorter but it also gets a bad performance. The result is 

![2019-10-31 10-15-53 的屏幕截图](https://user-images.githubusercontent.com/38291486/67913917-8e8b4180-fbc9-11e9-9966-ca3ee1d19af0.png)

For all experiments, I do not use the language model.

Thanks a lot!

**Comment by tomkocse** [comment]
@sw005320 Do you have a workable pure attention / CTC setup on ESPNET so that we can testify the performance of pure attention and pure CTC separately ?

**Comment by sw005320** [comment]
I have several invesitgation with RNN voxforge tasks (e.g., https://github.com/espnet/espnet/blob/master/egs/voxforge/asr1/RESULTS.md), but not transformer.

**Comment by tomkocse** [comment]
@sw005320 In your investigation (https://github.com/espnet/espnet/blob/master/egs/voxforge/asr1/RESULTS.md#scheduled-sampling-experiments-by-enabling-mtlalpha00-and-scheduled-sampling-ratio-with-00-and-05), it seems that the result is poor at the very beginning ( WER 74.8). Then the result becomes better when you change the scheduled-sampling-ratio from 0.0 to 0.5 (WER 15.0 ). 
However, the results become poor again when you are using Number of decoder layers = 2 (WER 73.2, 72.8), and the scheduled-sampling-ratio has no effect.

It seems that you are also getting poor results with pure attention mode, can you suggest how to solve the problem ?

**Comment by sw005320** [comment]
I have several experiences:
- I don't usually use more than one decoder layer. This would not significantly improve the performance and make training difficult (more data would solve it). 
- the number of encoder layers should also be smaller when the amount of training data is small.
- we may need some tweaks for neural network hyperparameters (initial learning rate, decay factor, decay strategy (scheduling), initialization). Again large training data would solve these issues.

If the amount of training data is small (e.g., <100), pure attention is unstable and we need some tweak (the above) to make it work.
https://arxiv.org/pdf/1811.02770.pdf describes a way to make pure-attention work in ESPnet.

@creatorscan, could you give me some comments? Also, can you commit your voxforge/WSJ pure attention config files to ours?

**Comment by Yingzhu-Zhao** [comment]
Is there a viable solution so far for pure attention transformer model on Librispeech dataset?

**Comment by tomkocse** [comment]
@sw005320 It will be great if you can release a fine-tuned pure LAS setup on Librispeech ? I think most ESPNET example are only tuned for mix CTC+LAS setup.

**Comment by RSSharma246** [comment]
Can the config files for pure attention mode/pure ctc be shared for the Voxforge results?
Thanks

**Comment by b-flo** [comment]
Looking at `egs/voxforge/asr1/RESULTS.md`, the"default" config for CTC should be the same as `egs/voxforge/conf/tuning/train_rnn.yaml` with `mtlalpha: 0.0` and `sampling-probability: 0.5`

**Comment by RSSharma246** [comment]
For attention only, alpha=0, and maxlenratio=0.8 and minlenratio=0.3 for both training and testing?
